### PR TITLE
feat: create dashboard for event analysis

### PR DESCRIPTION
**Pull request by llmin**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/lambda/api/router/reporting.ts [hunk 1]
```diff
@@ -19,13 +19,21 @@ const reporting_project = express.Router();
 const reportingServ: ReportingServ = new ReportingServ();
 
 reporting_project.post(
-  '',
+  '/funnel',
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/router/reporting.ts:22]: is it sankey or funnel visual?

File: src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts [hunk 2]
```diff
@@ -71,6 +70,11 @@ export interface DashboardCreateInputParameters {
   readonly pipelineId: string;
 }
 
+export interface VisualMapProps {
+  name: 'CHART' | 'TABLE' ;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts:74]: add `readonly` modifier?

File: src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts [hunk 3]
```diff
@@ -79,7 +83,25 @@ export interface CreateDashboardResult {
...
+}
+
+export interface VisualRelatedDefProps {
+  readonly timeScopeType: 'FIXED' | 'RELATIVE';
+  readonly sheetId: string;
+  readonly visualId: string;
+  readonly viewName: string;
+  readonly lastN?: number;
+  readonly timeUnit?: string;
+  readonly timeStart?: string;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts:103]: why not using `Date` type?

File: src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts [hunk 4]
```diff
@@ -79,7 +83,25 @@ export interface CreateDashboardResult {
...
+  readonly columnConfigurations?: FilterGroup;
+}
+
+export interface VisualRelatedDefProps {
+  readonly timeScopeType: 'FIXED' | 'RELATIVE';
+  readonly sheetId: string;
+  readonly visualId: string;
+  readonly viewName: string;
+  readonly lastN?: number;
+  readonly timeUnit?: string;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts:102]: how about defining it as `enum`?

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 5]
```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
...
+      visuals: visualPropsArray,
+      dashboardDef: dashboardDef as DashboardVersionDefinition,
+    });
+
+    logger.info(`final dashboard def: ${JSON.stringify(dashboard)}`);
+
+    let result: CreateDashboardResult;
+    if (!query.dashboardId) {
+
+      //crate QuickSight analysis
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:404]: typo 'crate'

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 6]
```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
...
+            'quicksight:RestoreAnalysis',
+            'quicksight:DeleteAnalysis',
+            'quicksight:UpdateAnalysisPermissions',
+            'quicksight:DescribeAnalysisPermissions',
+          ],
+        }],
+        Definition: dashboard as AnalysisDefinition,
+      });
+
+      //crate QuickSight dashboard
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:425]: typo

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 7]
```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
...
+        dashboardArn: newDashboard.Arn!,
+        dashboardName: `dashboard-${resourceName}`,
+        dashboardVersion: Number.parseInt(newDashboard.VersionArn!.substring(newDashboard.VersionArn!.lastIndexOf('/') + 1)),
+        analysisId,
+        analysisArn: newAnalysis.Arn!,
+        analysisName: `analysis-${resourceName}`,
+        visualIds: [],
+      };
+    } else {
+      //crate QuickSight analysis
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:458]: crate ->  update

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 8]
```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
...
+      const newAnalysis = await quickSight.updateAnalysis({
+        AwsAccountId: awsAccountId,
+        AnalysisId: query.analysisId,
+        Name: query.analysisName,
+        Definition: dashboard as AnalysisDefinition,
+      });
+
+      //crate QuickSight dashboard
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:466]: crate -> update

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 9]
```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
...
+      for (const _i of Array(60).keys()) {
+        try {
+          await quickSight.updateDashboardPublishedVersion({
+            AwsAccountId: awsAccountId,
+            DashboardId: query.dashboardId,
+            VersionNumber: Number.parseInt(versionNumber!),
+          });
+
+        } catch (err: any) {
+          if (err instanceof ConflictException ) {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:484]: Throw the error again if it's not `ConflictException` or retries are exhausted

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 10]
```diff
@@ -249,131 +193,315 @@ export class ReportingServ {
...
+        id: tableVisualId,
+      });
+
       return res.status(201).json(new ApiSuccess(result));
     } catch (error) {
       next(error);
     }
   };
 
+  private async create(resourceName: string, query: any, sqls: string[], datasetPropsArray: DataSetProps[], visualPropsArray: VisualProps[]) {
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:323]: It's a very long procedure for creating or updating the dashboard, how do you handle the error if it happens in any intermittent step?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts (hunk 2 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts:74]: fixed

File: src/control-plane/backend/lambda/api/router/reporting.ts (hunk 1 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/router/reporting.ts:22]: it funnel visual.  
sankey visual is used for path analysis.

plan to use 
/funnel  
/event
/path
/retention 
for our 4 type exploration dashboards.

File: src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts (hunk 4 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/service/quicksight/reporting-utils.ts:102]: fixed

File: src/control-plane/backend/lambda/api/service/reporting.ts (hunk 5 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/service/reporting.ts:404]: fixed

File: src/control-plane/backend/lambda/api/service/reporting.ts (hunk 10 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/service/reporting.ts:323]: will deal with this problem later, #149

**Review by tyyzqmf — APPROVED** [review]

**Review by zxkane — APPROVED** [review]

**Review by zxkane — CHANGES_REQUESTED** [review]
resolve conflicts locally

File: src/control-plane/backend/lambda/api/service/quicksight/sql-builder.ts [hunk 11]
```diff
@@ -38,8 +38,8 @@ export interface FunnelSQLParameters {
   readonly firstEventExtraCondition?: EventAndCondition;
   readonly eventAndConditions: EventAndCondition[];
   readonly timeScopeType: ExploreTimeScopeType;
-  readonly timeStart?: string;
-  readonly timeEnd?: string;
+  readonly timeStart?: Date;
+  readonly timeEnd?: Date;
```
Comment (tyyzqmf) [inline src/control-plane/backend/lambda/api/service/quicksight/sql-builder.ts:42]: Pass date as **string** from client side and convert it into the date from server side is the best approach.

**clientside:**
```
{
"time" : "2023-08-01"
}
```

**Server side:**
`const date = new Date(req.body.time)`

Comment (llmin) [reply src/control-plane/backend/lambda/api/service/quicksight/sql-builder.ts:42]: Does client side code using this interface?

Comment (llmin) [reply src/control-plane/backend/lambda/api/service/quicksight/sql-builder.ts:42]: fixed

**Comment by llmin** [comment]
>

**Review by zxkane — APPROVED** [review]

**Pull request merged**